modules/rostests/apitests/shell32/SHLimitInputEdit.cpp

```diff
@@ -0,0 +1,332 @@
+{
+public:
+    CShellFolder(INT iMaxNameLen = 0, BOOL bDisabled = FALSE)
+        : m_nRefCount(1)
+        , m_iMaxNameLen(iMaxNameLen)
+        , m_bDisabled(bDisabled)
+    {
+        trace("CShellFolder\n");
+    }
+
+    virtual ~CShellFolder()
+    {
+    }
+
+    /*** IUnknown methods ***/
+
+    STDMETHODIMP QueryInterface(REFIID riid, void **ppvObject);
+    STDMETHODIMP_(ULONG) AddRef();
+    STDMETHODIMP_(ULONG) Release();
+
+    /*** IShellFolder methods ***/
+
+    STDMETHODIMP ParseDisplayName(
+        HWND hwnd,
+        IBindCtx *pbc,
+        LPWSTR pszDisplayName,
+        ULONG *pchEaten,
+        PIDLIST_RELATIVE *ppidl,
+        ULONG *pdwAttributes)
+    {
+        return E_NOTIMPL;
+    }
+
+    STDMETHODIMP EnumObjects(
+        HWND hwnd,
+        SHCONTF grfFlags,
+        IEnumIDList **ppenumIDList)
+    {
+        return E_NOTIMPL;
+    }
+
+    STDMETHODIMP BindToObject(
+        PCUIDLIST_RELATIVE pidl,
+        IBindCtx *pbc,
+        REFIID riid,
+        void **ppv)
+    {
+        return E_NOTIMPL;
+    }
+
+    STDMETHODIMP BindToStorage(
+        PCUIDLIST_RELATIVE pidl,
+        IBindCtx *pbc,
+        REFIID riid,
+        void **ppv)
+    {
+        return E_NOTIMPL;
+    }
+
+    STDMETHODIMP CompareIDs(
+        LPARAM lParam,
+        PCUIDLIST_RELATIVE pidl1,
+        PCUIDLIST_RELATIVE pidl2)
+    {
+        return E_NOTIMPL;
+    }
+
+    STDMETHODIMP CreateViewObject(
+        HWND hwndOwner,
+        REFIID riid,
+        void **ppv)
+    {
+        return E_NOTIMPL;
+    }
+
+    STDMETHODIMP GetAttributesOf(
+        UINT cidl,
+        PCUITEMID_CHILD_ARRAY apidl,
+        SFGAOF *rgfInOut)
+    {
+        return E_NOTIMPL;
+    }
+
+    STDMETHODIMP GetUIObjectOf(
+        HWND hwndOwner,
+        UINT cidl,
+        PCUITEMID_CHILD_ARRAY apidl,
+        REFIID riid,
+        UINT *rgfReserved,
+        void **ppv)
+    {
+        return E_NOTIMPL;
+    }
+
+    STDMETHODIMP GetDisplayNameOf(
+        PCUITEMID_CHILD pidl,
+        SHGDNF uFlags,
+        STRRET *pName)
+    {
+        return E_NOTIMPL;
+    }
+
+    STDMETHODIMP SetNameOf(
+        HWND hwnd,
+        PCUITEMID_CHILD pidl,
+        LPCWSTR pszName,
+        SHGDNF uFlags,
+        PITEMID_CHILD *ppidlOut)
+    {
+        return E_NOTIMPL;
+    }
+
+    /*** IItemNameLimits methods ***/
+
+    STDMETHODIMP GetMaxLength(
+        LPCWSTR pszName,
+        int *piMaxNameLen);
+
+    STDMETHODIMP GetValidCharacters(
+        LPWSTR *ppwszValidChars,
+        LPWSTR *ppwszInvalidChars);
+
+protected:
+    LONG m_nRefCount;
+    INT m_iMaxNameLen;
+    BOOL m_bDisabled;
+};
+
+STDMETHODIMP CShellFolder::QueryInterface(REFIID riid, void **ppvObject)
+{
+    trace("QueryInterface\n");
+
+    if (!ppvObject)
+        return E_POINTER;
+
+    if (IsEqualIID(riid, IID_IItemNameLimits))
+    {
+        trace("IID_IItemNameLimits\n");
+        *ppvObject = static_cast<IItemNameLimits *>(this);
+        return S_OK;
```

Comment:
A succesful match in `QueryInterface` should call `AddRef`

Reply:
OK.